### ISSUE TITLE
Is Babelfish a Migration Tool?

### ISSUE THREAD

**Issue by btzq** [babelfish-for-postgresql/babelfish_extensions#2403]
Hi Community, 

We are using MSSQL right now and would love to move to Posgresql. (MSSQL is expensive). But, we have lots of Stored Procedures which is what is stopping us from migrating to Postgresql. 

Im trying to understand how Babelfish works, but it sounds to me like its either:
- A migration tool to convert entire MSSQL DB (including stored procs and others) into Postgresql equivalent? OR
- A connector on top of Postgresql? Meaning we need to deploy both MSSQL and Postgresql (with Babelfish plugin)?

Someone help me wrap my head around this in layman terms. Also, is Babelfish stable for production? Or is it still in development?

**Comment by robverschoor** [comment]
It is neither, although #2 comes closest. Babelfish is a Postgres-native implementation of (i) T-SQL syntax and semantics and (ii) the TDS protocol, in the form of Postgres extensions (it is really part of Postgres, not some additional external component). Babelfish implements T-SQL using Postgres building blocks; it does not 'translate' T-SQL queries into PG queries but -using its own T-SQL parser and QP extensions- generates what are essentially internal PG query trees that implement the T-SQL syntax and semantics. That is why 'connector' is much too simple a name.
With Babelfish, the PG instance also provides a TDS port (1433); when connecting to this port, the connection gets the T-SQL dialect and sees a SQL Server-like environment with a **master** and **tempdb** database and you can run most T-SQL unchanged.
For a current SQL Server-based app, the idea is to re-run the scripts that create the databases, tables, procedures, triggers etc. against Babelfish, then migrate the data (with DMS for example) and then just reconfigure the clients from connecting to the SQL Server port to the Babelfish TDS port.
It is recommended to run the [Compass too](https://github.com/babelfish-for-postgresql/babelfish_compass/releases/latest)l first to identify any aspects of T-SQL which Babelfish does not currently support.

**Comment by btzq** [comment]
I see, thanks for the information. 
I was really hoping it was the first one. Cause that would be a magical way for us to move out of MSSQL. 

I dont suppose you know of any tools in the market that can do that?

**Comment by robverschoor** [comment]
AWS provides the SCT tool to convert between SQL dialects: https://aws.amazon.com/dms/schema-conversion-tool/

**Comment by btzq** [comment]
Got it, thank you
